src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    public void stop() {
+        try {
+            OTExecutors.shutdownAndAwaitTermination(exec, Duration.ofSeconds(5));
+        } catch (final InterruptedException e) {
+            LOG.error("worker thread shutdown interrupted", e);
+            Thread.currentThread().interrupt();
+        }
+        monitor.close();
+    }
+
+    @Override
+    public void close() {
+        stop();
+    }
+
+    private void poll() {
+        topics.forEach(topic -> {
+            try {
+                pollUnsafe(topic);
+            } catch (final Exception e) {
+                LOG.error("caught exception for topic {}", topic, e);
+            }
+        });
+    }
+
+    private void pollUnsafe(final String topic) {
+        final Map<Integer, Long> sizes = monitor.getTopicSizes(topic);
+        sizes.forEach((part, size) -> gauge(String.format("%s.partition.%d.size", topic, part)).set(size));
+        gauge(String.format("%s.total.size", topic)).set(sumValues(sizes));
+
+        Map<Integer, Long> offsets = monitor.getGroupOffsets(groupId, topic);
+        final Map<Integer, Long> lag;
+        if (offsets.isEmpty()) {
+            // Consumer may not be consuming this topic yet (or consumer might not exist).
+            // In case consumer existed previously, we set all offsets and lag to 0.
+            offsets = lag = sizes
+                    .keySet()
+                    .stream()
+                    .collect(
+                            Collectors.toMap(
+                                    part -> part,
+                                    part -> 0L
+                            )
+                    );
+        } else {
+            if (!sizes.keySet().equals(offsets.keySet())) {
+                LOG.error("sizes/offsets partitions do not match for topic {}: {}/{}",
+                        topic, sizes.keySet(), offsets.keySet());
+                return;
+            }
+            final Map<Integer, Long> finalOffsets = offsets;
+            lag = sizes
+                    .keySet()
+                    .stream()
+                    .collect(
+                            Collectors.toMap(
+                                    part -> part,
+                                    part -> sizes.get(part) - finalOffsets.get(part)
+                            )
+                    );
+        }
+
+        offsets.forEach((part, off) -> gauge(String.format("%s.partition.%d.offset", topic, part)).set(off));
+        lag    .forEach((part, off) -> gauge(String.format("%s.partition.%d.lag",    topic, part)).set(off));
+
+        gauge(String.format("%s.total.offset", topic)).set(sumValues(offsets));
+        gauge(String.format("%s.total.lag",    topic)).set(sumValues(lag));
+    }
+
+    private AtomicLongGauge gauge(final String name) {
+        return (AtomicLongGauge) metricMap.get(name);
+    }
+
+    private static long sumValues(final Map<?, Long> map) {
+        return map.values().stream().mapToLong(size -> size).sum();
```

Comment:
Note that `Function.identity()` exists and would remove the lambda

Reply:
This doesn't work.  `mapToLong` takes a `ToLongFunction`, and `Function<T, T>` doesn't map to this easily.  This reads more cleanly than the function reference.  : [